compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp

```diff
@@ -49,6 +52,79 @@ struct HostDefaults {
   }
 };
 
+void ResolveCPUAndCPUFeatures(llvm::StringRef inCpu,
+                              llvm::StringRef inCpuFeatures,
+                              const llvm::Triple &triple, std::string &outCpu,
+                              std::string &outCpuFeatures) {
+  // Resolve "host"
+  if (inCpu == "host" || inCpuFeatures == "host") {
+    // If either Cpu or CpuFeatures is "host", the other must be either also
+    // host or the default value.
+    if (!inCpu.empty() && inCpu != "generic" && !inCpuFeatures.empty() &&
+        inCpuFeatures != "host") {
+      llvm::report_fatal_error(
+          "If either Cpu or CpuFeatures is `host`, the other must be either "
+          "also `host` or the default value.",
+          false);
+    }
+    const HostDefaults &hostDefaults = HostDefaults::get();
+    outCpu = hostDefaults.cpu;
+    outCpuFeatures = hostDefaults.cpuFeatures;
+  } else {
+    outCpu = inCpu;
+    outCpuFeatures = inCpuFeatures;
+  }
+
+  // Target-specific CPU feature tweaks that we need unconditionally.
+  if (triple.isAArch64()) {
+    llvm::SubtargetFeatures targetCpuFeatures(outCpuFeatures);
+    // x18 is platform-reserved per the Aarch64 procedure call specification.
+    targetCpuFeatures.AddFeature("reserve-x18", true);
+    outCpuFeatures = targetCpuFeatures.getString();
+  }
+
+  // If CPU is non-host and non-generic then we need to populate the
+  // corresponding features.
+  if (inCpu == "host" || inCpu == "generic" || inCpu.starts_with("generic-")) {
+    return;
+  }
+  if (triple.isX86()) {
+    llvm::SubtargetFeatures targetCpuFeatures(outCpuFeatures);
+    llvm::SmallVector<llvm::StringRef> cpuFeatureList;
+    llvm::X86::getFeaturesForCPU(outCpu, cpuFeatureList);
+    for (auto &feature : cpuFeatureList) {
+      targetCpuFeatures.AddFeature(feature);
+    }
+    outCpuFeatures = targetCpuFeatures.getString();
+    return;
```

Comment:
the early returns make this a bit hard to read - can you make this an if-else set to show the code paths are exclusive or make dedicated helper functions (`if (triple.isX86()) resolveCPUAndFeaturesForX86(...) else if (triple.isAarch64()) resolveCPUAndFeaturesForAarch64(...)`)